### ISSUE TITLE
Inheritance from intermediate traits fail at link time

### ISSUE THREAD

**Issue by jonas** [scala-native/scala-native#801]
The following test (and attached reachability test) 

tries to model the complex inheritance of the Java collections library. Compared to what is currently in `javalib`, it defines `trait Set[E]` as extending `Collection[E]` without any overrides. 

```scala
trait Collection[E] {
  def isEmpty: Boolean
  def size: Int
}
trait Set[E] extends Collection[E]

abstract class AbstractCollection[E] extends Collection[E] {
  override def isEmpty = size == 0
}
abstract class AbstractSet[E] extends AbstractCollection[E] with Set[E]

class TreeSet[E] extends AbstractSet[E] {
  override def size = 42
}

object Test {
  def main(args: Array[String]): Unit = {
    println((new TreeSet[Int]).size)
  }
}
```

Compiling it causes the following link error:

```
> sandbox/run
[info] Compiling 1 Scala source to /opt/scala-native/sandbox/target/scala-2.11/classes...
[info] Linking (596 ms)
[error] cannot link: @java.util.Set::add_java.lang.Object_bool
[error] cannot link: @java.util.Set::clear_unit
[error] cannot link: @java.util.Set::contains_java.lang.Object_bool
[error] cannot link: @java.util.Set::isEmpty_bool
[error] cannot link: @java.util.Set::iterator_java.util.Iterator
[error] cannot link: @java.util.Set::remove_java.lang.Object_bool
[error] cannot link: @java.util.Set::size_i32
[error] unable to link
[error] (sandbox/compile:nativeLinkNIR) unable to link
[error] Total time: 1 s, completed 9-Jun-2017 9:59:11 AM
```
[CollectionReachabilitySuite.scala.txt](https://github.com/scala-native/scala-native/files/1064576/CollectionReachabilitySuite.scala.txt)

**Comment by densh** [comment]
Might be a duplicate of #375.

**Comment by densh** [comment]
Thanks for the sample reachability test, it's going to speed up the reproduction quite a bit!

**Comment by jonas** [comment]
Yes, indeed looks like a dup of #375

**Comment by densh** [comment]
Fixed in #1323
